posts/scala-for-the-arrogant-part-2.md

```diff
@@ -0,0 +1,427 @@
+
+- Our semicolon key gets a rest as the majority of our functions are one-liners. Our method names are descriptive enough that composing them together adds readability to our solution. Admittedly, it can be more difficult to debug, but I would argue that structuring your programs in this manner creates a final result that contains far fewer defects than traditional methods.
+
+One critique you may have is that there is still quite a bit of verbosity to the code. For example, having a method that returns a Predicate can look unwieldy- especially if you are not familiar with that type of syntax. We also manage a parameterized List of Lists of Actions that we define as an Action Trail. We also manage our history as a Set of ArrayLists of Integers (we opt for ArrayList over generic List just to enforce the idea that the sequence of elements is important).
+
+# Refactoring Our Solution To Make It Scala
+Finally, we can rewrite this solution to use Scala. We will still retain the structure that we devised earlier, but let's see if the Scala syntax helps us create something more elegant. The inspiration for this solution comes from Martin Odersky's *[Functional Program Design In Scala](https://www.coursera.org/learn/progfun2)*, with some modifications made for clarity and to illustrate some different concepts.
+
+## One File
+Scala has a lot of nice features for development like worksheets: an interactive mode where you can see the results of your coding in real-time. We can also define a lot of our landscape in a single file. We will take that approach here. It may not be best practice but it will allow us to keep it simple and hold our solution in one space.
+
+## Define our own "Primitives"
+In our Java solution we continuously referred to our state by a list of integers (with each integer representing the volume of water currently in a jug) and our jugs as a list of integers (with each integer representing the capacity of the jug). Passing around two lists of ints with different meanings could be confusing and cause error. One alternative is to define new types that shadow the actual data underneath:
+```scala
+object WaterPouringTypes {
+  type Jug = Int
+  type State = List[Int]
+  type ActionTrail = List[List[Action]]
+}
+```
+Now we can pass a `Jug` as a parameter, have functions explicitly request a jug instead of an int, still have our logic treat it as an Int. For example, we'll take the Move Interface from our Java solution and model it as a trait, using the `Jug` and `State` types we just defined:
+
+```scala
+trait Move {
+    def doMove(state: State): State
+}
+case class Empty(jugIndex: Jug) extends Move {
+    override def doMove(state: State): State = state updated (jugIndex, 0)
```

Comment:
It is generally recommended to avoid using infix notation, see https://docs.scala-lang.org/style/method-invocation.html

Reply:
resolved